lightning/src/blinded_path/message.rs

```diff
@@ -74,16 +68,106 @@ impl Writeable for ForwardTlvs {
 	}
 }
 
+/// Similar to [`ForwardTlvs`], but these TLVs are for the final node.
+pub(crate) struct ReceiveTlvs {
+	pub tlvs: Option<RecipientData>
+}
+
+impl ReceiveTlvs {
+	pub fn new(tlvs: Option<RecipientData>) -> Self {
+		ReceiveTlvs { tlvs }
+	}
+}
+
+/// Represents additional data appended along with the sent reply path.
+///
+/// This data can be utilized by the final recipient for further processing
+/// upon receiving it back.
+#[derive(Clone, Debug)]
+pub enum RecipientData {
+	/// Represents the data specific to [`OffersMessage`]
+	///
+	/// [`OffersMessage`]: crate::onion_message::offers::OffersMessage
+	OffersContext(OffersData),
+
+	/// Represents the data appended with Custom Onion Message.
+	CustomContext(Vec<u8>),
+}
+
+/// Contains the data specific to [`OffersMessage`]
+///
+/// [`OffersMessage`]: crate::onion_message::offers::OffersMessage
+#[derive(Clone, Debug)]
+pub struct OffersData {
+	/// Payment ID of the outbound BOLT12 payment.
+	pub payment_id: Option<PaymentId>
+}
+
+impl RecipientData {
+	/// Creates a new RecipientData::OffersContext instance.
+	pub fn new_for_offers(payment_id: Option<PaymentId>) -> Self {
+		RecipientData::OffersContext(OffersData {
+			payment_id,
+		})
+	}
```

Comment:
Let's drop this constructor since (1) we don't want users to be able to create this (2) we'll likely have other variations in the future, and (3) the struct's fields are `pub` so it can be constructed without one.

Reply:
Actually, we'll have to make the fields `pub(crate)` to make ensure (1).

Reply:
Removed the redundant constructor in [8e4e1aa](https://github.com/lightningdevkit/rust-lightning/commit/8e4e1aabadb5fda9f4d0ec2c845263c94c273643)

Regarding the visibility of the data, I think we'll need to keep it as `pub`. Since `RecipientData` is a field within `PeeledOnion`, which itself is a public struct, it seems like restricting the internal fields' visibility any further might not be possible.

```rust
pub enum PeeledOnion<T: OnionMessageContents> {
	///
	Receive(ParsedOnionMessageContents<T>, Option<RecipientData>, Option<BlindedPath>)
}```

Let me know what you think about it!

Reply:
Seems like making `payment_id` have `pub(crate)` visibility works. Hmmm... but I guess that would restrict who could use that to implement `OffersMessageHandler`. So maybe we'll need to keep it as you have it now afterall.